### PR TITLE
[docs] Apply the new DX to the Button Field demos

### PR DESCRIPTION
**Pull request by flaviendelangle**
Part of #14496

[Doc preview](https://deploy-preview-14860--material-ui-x.netlify.app/x/react-date-pickers/custom-field/#using-a-button)

## Packages

- [x] Introduce `DateRangePickerFieldProps` and `DateTimeRangePickerFieldProps`
- [x] Make sure every picker passes a format to the field, the date range ones did not (the `useDesktopPicker` and equivalent hooks to not enforce it yet because it was non trivial, once we have more info passed through the context it should be easier to enforce this kind of things, right now the typing is super hard to modify without introducing TS errors).

## Docs

- [x] Move each category of custom behavior demo in a folder to make the "custom-field" folder more readable
- [x] Add validation support on the demos
- [x] Add placeholder support on the demos
- [x] Stop relying on complex types and instead use the new `DatePickerFieldProps` pattern
- [x] Pass the forwarded props to the `Button` to support `sx`, `className` etc...
- [x] Improve the phrasing in the impacted doc section

I did not migrate those demo to use the new multi file demo feature because it is broken when used inside nested folder.

### CODE REVIEW

File: packages/x-date-pickers/src/internals/models/fields.ts
Comment (flaviendelangle) [inline packages/x-date-pickers/src/internals/models/fields.ts:13]: The format is always defined when a field is used inside a picker (I'm currently making sure it is, because it makes the customization easier) and so is the value

File: packages/x-date-pickers/src/internals/models/fields.ts
Comment (flaviendelangle) [inline packages/x-date-pickers/src/internals/models/fields.ts:21]: Since the field is used as a slot inside the picker, it receives an `ownerState`

And some (most?) `@mui/material` components don't like when you pass an `ownerState` to them (it overrides their own `ownerState`).
So I had to add it to the typing here to be able to not forward it to the `Button` in the updated demos

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (michelengelen) [inline docs/data/date-pickers/custom-field/custom-field.md:131]:
```suggestion
you can replace the field with the `<Autocomplete />` from `@mui/material` to list those dates:
```

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (michelengelen) [inline docs/data/date-pickers/custom-field/custom-field.md:135]: Not sure if this was intentional, but did you mean to use 'Text Field' or 'TextField' (like in the text below)?

```suggestion
### Using a read-only TextField
```

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (michelengelen) [inline docs/data/date-pickers/custom-field/custom-field.md:145]: same here

```suggestion
and you don't want the UI to look like a TextField, you can replace the field with a `<Button />` from `@mui/material`:
```

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (michelengelen) [reply docs/data/date-pickers/custom-field/custom-field.md:135]: or according to this doc: https://www.notion.so/mui-org/Style-conventions-for-MUI-components-and-products-dff7d2d7d1ba4a4abd47f48cf345b800?pvs=4#1279b9e04d514d32a109b846a0945813

shouldn't we use 'text field' instead?

Comment (flaviendelangle) [reply docs/data/date-pickers/custom-field/custom-field.md:135]: We are describing the `<TextField />` from `@mui/material` here, not the generic concept of a text field, so I don't think "text field" would be appropriated.

I'm using "Text Field" because it's a reference to one our our product (from the core team), similar to the Data Grid or the Date Time Picker.

I'll ping Sam for a validation of this PR before merging :+1:

Comment (michelengelen) [reply docs/data/date-pickers/custom-field/custom-field.md:135]: Well, technically it is not a product ... I guess... no? 🤔

Comment (samuelsycamore) [reply docs/data/date-pickers/custom-field/custom-field.md:135]: We're describing our specific implementation (rather than the general concept), so I would style it as a proper noun: Text Field.

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (samuelsycamore) [inline docs/data/date-pickers/custom-field/custom-field.md:131]: I would style this as a proper noun rather than two backticked elements.

```suggestion
you can replace the field with Material UI Autocomplete to list those dates:
```

Comment (samuelsycamore) [reply docs/data/date-pickers/custom-field/custom-field.md:131]: I'm curious if @vladmoroz has any thoughts about formatting in this context.

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (samuelsycamore) [inline docs/data/date-pickers/custom-field/custom-field.md:138]:
```suggestion
but you still want the UI to look like a Text Field, you can replace the field with a read-only Material UI Text Field component:
```

Comment (samuelsycamore) [reply docs/data/date-pickers/custom-field/custom-field.md:138]: I would also like to see the line break removed from this sentence—try to keep it to one sentence per line, even if it's a very long sentence.

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (samuelsycamore) [inline docs/data/date-pickers/custom-field/custom-field.md:145]: Here's how I would style this:

```suggestion
and you don't want the UI to look like a Text Field, you can replace the field with a Material UI Button:
```

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (samuelsycamore) [inline docs/data/date-pickers/custom-field/custom-field.md:149]: let's capitalize this since we're referring to the MUI X components.

```suggestion
The same logic can be applied to any Range Picker:
```

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (vladmoroz) [reply docs/data/date-pickers/custom-field/custom-field.md:131]: I'd say it like this personally...

> If your user can only select a value in a small list of available dates, you can replace the field with [Autocomplete](https://mui.com/material-ui/react-autocomplete/) to list those dates:

This sounds OK to me too:

> If your user can only select a value in a small list of available dates, you can replace the field with the Autocomplete component from `@mui/material` to list those dates:

- Proper noun either way
- I'd link out to Autocomplete because I need the API handy, and it saves us a reference to Material UI. Otherwise, my preference is explicit `@mui/material` instead of "Material UI" because people confuse MUI / Material UI / Material Design all the time.
- Either "with the Autocomplete component from..." or "with Autocomplete from...". Just "the Autocomplete" doesn't sound right to my ear.

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (flaviendelangle) [reply docs/data/date-pickers/custom-field/custom-field.md:138]: What is the goal of removing line break on the markdown file?
Those do not impact the line break on the doc..

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (flaviendelangle) [reply docs/data/date-pickers/custom-field/custom-field.md]: I updated the 3 sentences with a link to our `@mui/material` doc.

I just kept the "with the [Autocomplete]() component" because just "with [Autocomplete()" felt weird to me (same for "with [Button]().".

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (samuelsycamore) [reply docs/data/date-pickers/custom-field/custom-field.md]: It's just simpler to maintain one sentence per line—if I wanted to edit this sentence more heavily I'd have to leave two separate comments, or leave a suggestion across the two lines. We don't need to add our own line wrapping when an IDE will do that for us.

File: packages/x-date-pickers/src/internals/models/fields.ts
Comment (LukasTy) [reply packages/x-date-pickers/src/internals/models/fields.ts:21]: Do we want to update the type of it to `PickersOwnerState`? 🤔

File: docs/data/date-pickers/custom-field/behavior-button/MaterialDateRangePicker.tsx
Comment (LukasTy) [inline docs/data/date-pickers/custom-field/behavior-button/MaterialDateRangePicker.tsx:12]: WDYT, should we re-export `useValidation` from `@mui/x-date-pickers-pro/validation`? 🤔 
Would make at least this import cleaner. 🤷 
```suggestion
import { useValidation, validateDateRange } from '@mui/x-date-pickers-pro/validation';
```

File: docs/data/date-pickers/custom-field/behavior-button/MaterialDateRangePicker.tsx
Comment (LukasTy) [inline docs/data/date-pickers/custom-field/behavior-button/MaterialDateRangePicker.tsx:63]: Nitpick: Avoid abbreviation. 🙈 
```suggestion
  const formattedValue = (value ?? [null, null])
    .map((date) => (date == null ? parsedFormat : date.format(format)))
    .join(' – ');

  return (
    <Button
      {...other}
      variant="outlined"
      color={hasValidationError ? 'error' : 'primary'}
      ref={InputProps?.ref}
      onClick={handleTogglePicker}
    >
      {label ? `${label}: ${formattedValue}` : formattedValue}
```

File: docs/data/date-pickers/custom-field/behavior-button/MaterialDateRangePicker.tsx
Comment (LukasTy) [inline docs/data/date-pickers/custom-field/behavior-button/MaterialDateRangePicker.tsx:68]: WDYT, is it worth adding a `TODO v8` prefix?
```suggestion
// TODO v8: Will be removed before the end of the alpha since single input will become the default field.
```

File: docs/data/date-pickers/custom-field/behavior-button/MaterialDatePicker.tsx
Comment (LukasTy) [inline docs/data/date-pickers/custom-field/behavior-button/MaterialDatePicker.tsx:60]:
```suggestion
  const formattedValue = value == null ? parsedFormat : value.format(format);

  return (
    <Button
      {...other}
      variant="outlined"
      color={hasValidationError ? 'error' : 'primary'}
      ref={InputProps?.ref}
      onClick={handleTogglePicker}
    >
      {label ? `${label}: ${formattedValue}` : formattedValue}
```

**Review by LukasTy — APPROVED** [review]
Nice improvement over the existing approach, great work! 💙 👏

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (flaviendelangle) [reply docs/data/date-pickers/custom-field/custom-field.md]: > We don't need to add our own line wrapping when an IDE will do that for us.

Webstorm advises me to add those line breaks though :laughing:

Comment (samuelsycamore) [reply docs/data/date-pickers/custom-field/custom-field.md]: Interesting! I don't think we should listen to Webstorm in this case. We don't have a consistent rule to fall back on to decide when to add a line break like this, so I think it would make the docs look really messy if we started doing this everywhere. If we wanted to implement a style rule like this then I think it would need to happen programmatically via Vale or Prettier maybe. I've never used Webstorm before, though. Does it wrap lines for you, or does it display a sentence like this as one long line?

Comment (flaviendelangle) [reply docs/data/date-pickers/custom-field/custom-field.md]: > it would make the docs look really messy if we started doing this everywhere

I'm basically doing it everywhere
But if we say that we don't do line break inside a sentence, that's fine for me.

> Does it wrap lines for you, or does it display a sentence like this as one long line?

It just adds a visual feedback, similar to the one when I have a spelling error, I can probably disable it.

File: packages/x-date-pickers/src/internals/models/fields.ts
Comment (flaviendelangle) [reply packages/x-date-pickers/src/internals/models/fields.ts:21]: We need #14994

File: docs/data/date-pickers/custom-field/behavior-button/MaterialDateRangePicker.tsx
Comment (LukasTy) [reply docs/data/date-pickers/custom-field/behavior-button/MaterialDateRangePicker.tsx:12]: Gotcha. 👍 
Makes sense. 💯

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (samuelsycamore) [inline docs/data/date-pickers/custom-field/custom-field.md:137]:
```suggestion
but you still want the UI to look like a Text Field, you can replace the field with a read-only [Text Field](/material-ui/react-text-field/) component:
```

**Review by samuelsycamore — APPROVED** [review]
One last nitpick here but otherwise the doc changes look good. 👍

File: docs/data/date-pickers/custom-field/custom-field.md
Comment (samuelsycamore) [reply docs/data/date-pickers/custom-field/custom-field.md]: > But if we say that we don't do line break inside a sentence, that's fine for me.

Let's stick with that then, for the sake of consistency between authors and products.

**Pull request merged**
